### PR TITLE
Gho ccip bridge

### PR DESCRIPTION
**Pull request by LucasWongC**
### Overview
The `CcipGhoBridge` smart contract is a custom implementation designed to facilitate the bridging of GHO tokens between different blockchain networks using Chainlink's Cross-Chain Interoperability Protocol (CCIP). It enables token transfers between chains in a secure and decentralized manner by leveraging Chainlink's infrastructure. Here's an overview of its key features and components:

### Key Components
1. **Router, LINK, GHO Addresses**:
   - **`ROUTER`**: The Chainlink CCIP router address used for cross-chain communication.
   - **`LINK`**: Address of the Chainlink token (LINK), used for paying transaction fees when required.
   - **`GHO`**: Address of the GHO token, which the contract helps bridge across chains.

2. **Bridges Mapping**: 
   - A mapping of destination chain selectors (chain identifiers) to the corresponding bridge address on the destination chain.

### Main Functionalities
1. **Token Transfers**:
   - The contract allows users to transfer GHO tokens from one blockchain to another. It collects all GHO tokens from the sender, validates the transfer, and sends the tokens via Chainlink CCIP to the specified destination chain.
   - A modifier `checkDestination` ensures the destination chain and bridge are properly configured.

2. **Fee Payment**:
   - Users can pay fees in either LINK tokens or native gas tokens of the chain (e.g., ETH).
   - The contract calculates the required fees via the `IRouterClient.getFee` method and handles fee payments accordingly.

3. **Cross-Chain Message Handling**:
   - When a cross-chain message is received, the `_ccipReceive` function decodes the message, verifies its authenticity, and releases the corresponding GHO tokens to the specified recipients on the destination chain.

4. **Quote Transfer**:
   - This function, `quoteTransfer`, allows users to estimate the fee required for a token transfer without actually performing the transfer.

5. **Setting Destination Bridges**:
   - The contract owner can configure or update the bridge addresses for different destination chains through the `setDestinationBridge` function.

### Security Features
- **Ownership Control**: The contract uses OpenZeppelin's `Ownable` contract to ensure only the owner can configure or update key components like destination bridges.
- **Validation**: It checks whether the destination chain and bridge address are valid before executing transfers, ensuring that users cannot send tokens to unsupported chains.

### Events
- **TransferIssued**: Emitted when a transfer is successfully initiated.
- **TransferFinished**: Emitted when a transfer completes successfully on the destination chain.
- **DestinationUpdated**: Emitted when a new bridge address is set for a specific chain.

### Error Handling
- **UnsupportedChain**: Thrown when attempting to transfer to an unsupported destination chain.
- **InvalidTransferAmount**: Thrown when the total amount of tokens to transfer is zero.
- **InsufficientFee**: Thrown if the provided fee is insufficient to cover the transaction.
- **InvalidMessage**: Thrown when the message received from a source chain does not come from the expected bridge address.

### Conclusion
The `CcipGhoBridge` smart contract provides a robust mechanism for bridging GHO tokens across multiple chains via Chainlink's CCIP. It offers flexibility in fee payments, supports secure cross-chain transfers, and allows for easy configuration of destination bridges.

### CODE REVIEW

File: scripts/DeployBridges.s.sol [hunk 1]
```diff
@@ -60,3 +61,27 @@ contract DeployArbBridgeArbitrum is ArbitrumScript {
...
+      0x3765A685a401622C060E5D700D9ad89413363a91
+    );
+  }
+}
+
+contract DeployCcipGhoBridgeArbitrum is ArbitrumScript {
+  function run() external broadcast {
+    bytes32 salt = 'GHO Chainlink CCIP Bridge';
+    new CcipGhoBridge{salt: salt}(
+      0x141fa059441E0ca23ce184B6A78bafD2A517DdE8,
```
Comment (efecarranza) [inline scripts/DeployBridges.s.sol:81]: What are these addresses?

File: scripts/DeployBridges.s.sol [hunk 2]
```diff
@@ -60,3 +61,27 @@ contract DeployArbBridgeArbitrum is ArbitrumScript {
     new AaveArbEthERC20Bridge{salt: salt}(0x3765A685a401622C060E5D700D9ad89413363a91);
   }
 }
+
+contract DeployCcipGhoBridgeEthereum is EthereumScript {
+  function run() external broadcast {
+    bytes32 salt = 'GHO Chainlink CCIP Bridge';
+    new CcipGhoBridge{salt: salt}(
+      0x80226fc0Ee2b096224EeAc085Bb9a8cba1146f7D,
```
Comment (efecarranza) [inline scripts/DeployBridges.s.sol:69]: What are these addresses?

File: scripts/DeployBridges.s.sol (hunk 1 above)
Comment (efecarranza) [reply scripts/DeployBridges.s.sol:81]: Please add a name to them or at least a link to Etherscan and a brief description.

File: src/bridges/chainlink-ccip/CcipGhoBridge.sol [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+contract CcipGhoBridge is ICcipGhoBridge, CCIPReceiver, Ownable {
+  /// @dev Chainlink CCIP router address
+  address public immutable ROUTER;
+  /// @dev LINK token address
+  address public immutable LINK;
+  /// @dev GHO token address
+  address public immutable GHO;
+
+  /// @dev Address of bridge (chainSelector => bridge address)
+  mapping(uint64 => address) public bridges;
```
Comment (efecarranza) [inline src/bridges/chainlink-ccip/CcipGhoBridge.sol:28]: You can name the variables:

`mapping (uint64 selector => address bridge) public bridges`

File: src/bridges/chainlink-ccip/CcipGhoBridge.sol [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+  /// @inheritdoc ICcipGhoBridge
+  function transfer(
+    uint64 destinationChainSelector,
+    Transfer[] calldata transfers,
+    PayFeesIn payFeesIn
+  ) external payable checkDestination(destinationChainSelector) returns (bytes32 messageId) {
+    uint256 totalAmount;
+
+    for (uint256 i; i < transfers.length; ) {
```
Comment (efecarranza) [inline src/bridges/chainlink-ccip/CcipGhoBridge.sol:62]: Let's cache the transfers.length in a variable to save some reads and gas.

File: src/bridges/chainlink-ccip/CcipGhoBridge.sol [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+import {Client} from '@chainlink/contracts-ccip/src/v0.8/ccip/libraries/Client.sol';
+
+import {ICcipGhoBridge} from './ICcipGhoBridge.sol';
+
+/**
+ * @title CcipGhoBridge
+ * @author LucasWongC
+ * @notice Helper contract to bridge GHO using Chainlink CCIP
+ */
+contract CcipGhoBridge is ICcipGhoBridge, CCIPReceiver, Ownable {
```
Comment (efecarranza) [inline src/bridges/chainlink-ccip/CcipGhoBridge.sol:19]: I'd call this `AaveCcip...`

File: src/bridges/chainlink-ccip/CcipGhoBridge.sol [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    }
+    _;
+  }
+
+  /// @inheritdoc ICcipGhoBridge
+  function transfer(
+    uint64 destinationChainSelector,
+    Transfer[] calldata transfers,
+    PayFeesIn payFeesIn
+  ) external payable checkDestination(destinationChainSelector) returns (bytes32 messageId) {
```
Comment (efecarranza) [inline src/bridges/chainlink-ccip/CcipGhoBridge.sol:59]: This function is callable by anyone, we should only allow it to be called by OwnerOrGuardian and add a guaridan role to it

File: src/bridges/chainlink-ccip/CcipGhoBridge.sol [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    bytes32 messageId = message.messageId;
+    uint64 sourceChainSelector = message.sourceChainSelector;
+    address bridge = abi.decode(message.sender, (address));
+    Transfer[] memory transfers = abi.decode(message.data, (Transfer[]));
+
+    if (bridge != bridges[sourceChainSelector]) {
+      revert InvalidMessage();
+    }
+
+    for (uint256 i; i < transfers.length; ) {
```
Comment (efecarranza) [inline src/bridges/chainlink-ccip/CcipGhoBridge.sol:153]: Cache length to save gas

File: src/bridges/chainlink-ccip/CcipGhoBridge.sol [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+      IERC20(GHO).transfer(transfers[i].to, transfers[i].amount);
```
Comment (efecarranza) [inline src/bridges/chainlink-ccip/CcipGhoBridge.sol:154]: We likely only want to transfer to the collector

File: tests/bridges/chainlink-ccip/CcipGhoBridgeTest.t.sol [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+    sourceFork = vm.createFork(vm.rpcUrl('mainnet'));
+
+    bob = makeAddr('bob');
+    alice = makeAddr('alice');
+
+    ccipLocalSimulatorFork = new CCIPLocalSimulatorFork();
+    vm.makePersistent(address(ccipLocalSimulatorFork));
+
+    Register.NetworkDetails memory destinationNetworkDetails = Register.NetworkDetails({
+      chainSelector: 4949039107694359620,
```
Comment (efecarranza) [inline tests/bridges/chainlink-ccip/CcipGhoBridgeTest.t.sol:48]: Please have these addresses as variables with links to Etherscan

**Review by efecarranza — CHANGES_REQUESTED** [review]

File: src/bridges/chainlink-ccip/CcipGhoBridge.sol (hunk 8 above)
Comment (LucasWongC) [reply src/bridges/chainlink-ccip/CcipGhoBridge.sol:154]: But I think generalization is good for the future

Comment (efecarranza) [reply src/bridges/chainlink-ccip/CcipGhoBridge.sol]: @LucasWongC Probably won't be approved by BGD

File: src/bridges/chainlink-ccip/AaveCcipGhoBridge.sol [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+    });
+
+    fee = IRouterClient(ROUTER).getFee(destinationChainSelector, message);
+  }
+
+  /// @inheritdoc CCIPReceiver
+  function _ccipReceive(Client.Any2EVMMessage memory message) internal override {
+    bytes32 messageId = message.messageId;
+    uint64 sourceChainSelector = message.sourceChainSelector;
+    address bridge = abi.decode(message.sender, (address));
```
Comment (Luigy-Lemon) [inline src/bridges/chainlink-ccip/AaveCcipGhoBridge.sol:145]:
```suggestion
``` 
It doesn't matter who sent the GHO

File: src/bridges/chainlink-ccip/AaveCcipGhoBridge.sol [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+    Client.EVMTokenAmount[] memory tokenAmounts = message.destTokenAmounts;
+
+    if (bridge != bridges[sourceChainSelector] || tokenAmounts.length != 1) {
```
Comment (Luigy-Lemon) [inline src/bridges/chainlink-ccip/AaveCcipGhoBridge.sol:148]:
```suggestion
    if (tokenAmounts[0].token != GHO || tokenAmounts[0].amount == 0) {
```

It doesn't really matter who sent the GHO to AAVE in a new chain. Enforcing the sender being the corresponding bridge on other network introduces an unnecessary restriction.

The messageId will give all the information needed. All that needs to be checked is if GHO has been sent or not.

File: src/bridges/chainlink-ccip/AaveCcipGhoBridge.sol [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+      feeToken: payFeesIn == PayFeesIn.LINK ? LINK : address(0)
+    });
+
+    fee = IRouterClient(ROUTER).getFee(destinationChainSelector, message);
+  }
+
+  /// @inheritdoc CCIPReceiver
+  function _ccipReceive(Client.Any2EVMMessage memory message) internal override {
+    bytes32 messageId = message.messageId;
+    uint64 sourceChainSelector = message.sourceChainSelector;
```
Comment (Luigy-Lemon) [inline src/bridges/chainlink-ccip/AaveCcipGhoBridge.sol:144]:
```suggestion
```
It doesn't matter where it comes from

**Review by Luigy-Lemon — COMMENTED** [review]
I don't consider this a complete review as it would require a deeper look into how CCIP works and playing with it. 

I believe the ccipReceive is overly strict as it stands and could impose various issues and delays to missed configurations, alternative sources of funds, or direct interactions of the Aave Executor with this contract at the time of sending.  
The msg.sender cached in the messageId may differ from the bridge contract.

**Review by efecarranza — APPROVED** [review]

**Pull request merged**